homeassistant/components/sma/config_flow.py

```diff
@@ -51,34 +69,51 @@ def __init__(self) -> None:
             CONF_GROUP: GROUPS[0],
             CONF_PASSWORD: vol.UNDEFINED,
         }
+        self._discovery_data: dict[str, Any] = {}
+
+    async def _handle_user_input(
+        self, user_input: dict[str, Any], discovery: bool = False
+    ) -> tuple[dict[str, str], dict[str, str]]:
+        """Handle the user input."""
+        errors: dict[str, str] = {}
+        device_info: dict[str, str] = {}
+
+        if not discovery:
+            self._data[CONF_HOST] = user_input[CONF_HOST]
+
+        self._data[CONF_SSL] = user_input[CONF_SSL]
+        self._data[CONF_VERIFY_SSL] = user_input[CONF_VERIFY_SSL]
+        self._data[CONF_GROUP] = user_input[CONF_GROUP]
+        self._data[CONF_PASSWORD] = user_input[CONF_PASSWORD]
+
+        try:
+            device_info = await validate_input(
+                self.hass, user_input=user_input, data=self._data
+            )
+        except pysma.exceptions.SmaConnectionException:
+            errors["base"] = "cannot_connect"
+        except pysma.exceptions.SmaAuthenticationException:
+            errors["base"] = "invalid_auth"
+        except pysma.exceptions.SmaReadException:
+            errors["base"] = "cannot_retrieve_device_info"
+        except Exception:
+            _LOGGER.exception("Unexpected exception")
+            errors["base"] = "unknown"
+
+        return errors, device_info
 
     async def async_step_user(
         self, user_input: dict[str, Any] | None = None
     ) -> ConfigFlowResult:
         """First step in config flow."""
-        errors = {}
+        errors: dict[str, str] = {}
         if user_input is not None:
-            self._data[CONF_HOST] = user_input[CONF_HOST]
-            self._data[CONF_SSL] = user_input[CONF_SSL]
-            self._data[CONF_VERIFY_SSL] = user_input[CONF_VERIFY_SSL]
-            self._data[CONF_GROUP] = user_input[CONF_GROUP]
-            self._data[CONF_PASSWORD] = user_input[CONF_PASSWORD]
-
-            try:
-                device_info = await validate_input(self.hass, user_input)
-            except pysma.exceptions.SmaConnectionException:
-                errors["base"] = "cannot_connect"
-            except pysma.exceptions.SmaAuthenticationException:
-                errors["base"] = "invalid_auth"
-            except pysma.exceptions.SmaReadException:
-                errors["base"] = "cannot_retrieve_device_info"
-            except Exception:
-                _LOGGER.exception("Unexpected exception")
-                errors["base"] = "unknown"
+            errors, device_info = await self._handle_user_input(user_input=user_input)
 
             if not errors:
                 await self.async_set_unique_id(str(device_info["serial"]))
```

Comment:
```suggestion
                await self.async_set_unique_id(str(device_info["serial"]), raise_on_progress=False)
```

Reply:
Otherwise if the device is discovered and someone tries to start a second flow (via the user step) to connect with the same device (with the same unique id) it would raise, with this change we prefer the user flow over the discovery one in that scenario.